x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,204 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cometd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/filebeat/channel"
+	"github.com/elastic/beats/v7/filebeat/input"
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/common/transport/httpcommon"
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	bay "github.com/elastic/bayeux"
+)
+
+const (
+	inputName = "cometd"
+)
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *cometdInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Input worker has started.")
+			defer in.log.Info("Input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+func (in *cometdInput) run() error {
+	b := bay.Bayeux{}
+	creds := bay.GetSalesforceCredentials()
+	in.out = b.Channel(in.out, "-1", creds, in.config.ChannelName)
+
+	var event Event
+	for e := range in.out {
+		if !e.Successful {
+			// To handle the last response where the object received was empty
+			if e.Data.Payload == nil {
+				return nil
+			}
+
+			// Convert json.RawMessage response to []byte
+			msg, err := json.Marshal(e.Data.Payload)
```

Comment:
Is this call necessary? `json.RawMessage` is just an alias for an already JSON encoded `[]byte`.

It has a method to do this conversion: https://pkg.go.dev/encoding/json#RawMessage.MarshalJSON

(Really it is just a type cast: https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/encoding/json/stream.go;l=263)

Reply:
Understood, thank you.